### PR TITLE
Store FileReview records per reviewed file 

### PR DESCRIPTION
**Pull request by gylaz**
Violations now belong to FileReview record,
and builds have many file reviews.

File reviews have a completed status which can be used for jobs.

### CODE REVIEW

File: app/models/style_guide/coffee_script.rb [hunk 1]
```diff
@@ -4,18 +4,18 @@ class CoffeeScript < Base
...
-
-        Violation.new(
-          filename: file.filename,
-          line: line,
-          patch_position: line.patch_position,
-          line_number: violation["lineNumber"],
-          messages: [violation["message"]]
-        )
+    def file_review(file)
+      FileReview.new(filename: file.filename).tap do |file_review|
```
Comment (thorncp) [inline app/models/style_guide/coffee_script.rb]: Don't need `tap` here, `new` will accept the block.

File: app/models/style_guide/coffee_script.rb [hunk 2]
```diff
@@ -4,18 +4,18 @@ class CoffeeScript < Base
...
+        content = content_for_file(file)
+        lint(content).map do |violation|
```
Comment (thorncp) [inline app/models/style_guide/coffee_script.rb]: `each` instead of `map` since we're saving the return value?

File: app/models/violations.rb [hunk 3]
```diff
@@ -1,33 +0,0 @@
-# Container of unique violations
-class Violations
```
Comment (thorncp) [inline app/models/violations.rb]: :scissors: :+1:

File: db/migrate/20150526214750_associate_violations_with_file_reviews.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  private
+
+  def associate_violations_with_file_reviews
+    Violation.find_each do |violation|
+      file_review = FileReview.find_or_create_by!(
+        build_id: violation.build_id,
+        filename: violation.filename,
+        completed_at: Time.current
```
Comment (thorncp) [inline db/migrate/20150526214750_associate_violations_with_file_reviews.rb]: Seems like including this in the query will end up creating a file review for every violation.

Should we do something like this?

```ruby
attrs = { build_id: violation.build_id, filename: violation.filename }
file_review = FileReview.find_or_create_by!(attrs) do |file_review|
  file_review.completed_at = Time.current
end
```

File: app/models/style_guide/coffee_script.rb (hunk 2 above)
Comment (teoljungberg) [inline app/models/style_guide/coffee_script.rb]: I think it could be good to have `map` over `each` for debugging purposes.

File: app/models/style_guide/ruby.rb [hunk 5]
```diff
@@ -3,20 +3,16 @@ module StyleGuide
...
-          line = file.line_at(violation.line)
-
-          Violation.new(
-            filename: file.filename,
-            patch_position: line.patch_position,
-            line: line,
-            line_number: violation.line,
-            messages: [violation.message]
-          )
+        FileReview.new(filename: file.filename).tap do |file_review|
```
Comment (teoljungberg) [inline app/models/style_guide/ruby.rb]: Can this be extracted into a private method for the style guide classes? `conduct_file_review` or something along those lines

File: spec/models/file_review_spec.rb [hunk 6]
```diff
@@ -5,8 +5,47 @@
...
-    it { should validate_presence_of :build }
+  describe "#build_violation" do
+    context "when line has been changed" do
+      it "builds a violations" do
+        line = double("Line", changed?: true, patch_position: 121)
+        file_review = FileReview.new
+        expected_violation = Violation.new(
+          patch_position: 121,
+          line_number: 1,
+          messages: ["hello"]
```
Comment (teoljungberg) [inline spec/models/file_review_spec.rb]: How about a trailing comma?

File: spec/models/style_checker_spec.rb [hunk 7]
```diff
@@ -1,17 +1,18 @@
 require "rails_helper"
 
-describe StyleChecker, "#violations" do
-  it "returns a collection of computed violations" do
+describe StyleChecker, "#file_reviews" do
```
Comment (teoljungberg) [inline spec/models/style_checker_spec.rb]: How about having the `describe "#file_reviews"` on a separate line?

File: spec/models/style_guide/coffee_script_spec.rb [hunk 8]
```diff
@@ -31,26 +31,30 @@
     end
   end
 
-  describe "#violations_in_file" do
+  describe "#file_review" do
+    it "returns completed file review" do
+      repo_config = double("RepoConfig", enabled_for?: true, for: {})
+      style_guide = StyleGuide::CoffeeScript.new(repo_config, "bob")
+      file = build_file("foo")
```
Comment (teoljungberg) [inline spec/models/style_guide/coffee_script_spec.rb]: How about having the filename being a keyword argument for easier readability?

File: spec/models/style_guide/coffee_script_spec.rb [hunk 9]
```diff
@@ -31,26 +31,30 @@
     end
   end
 
-  describe "#violations_in_file" do
+  describe "#file_review" do
+    it "returns completed file review" do
+      repo_config = double("RepoConfig", enabled_for?: true, for: {})
```
Comment (teoljungberg) [inline spec/models/style_guide/coffee_script_spec.rb]: We're creating a lot of repo configs in this test, how about if we add a `build_repo_config` factory?

File: spec/models/style_guide/coffee_script_spec.rb (hunk 8 above)
Comment (teoljungberg) [inline spec/models/style_guide/coffee_script_spec.rb]: Oh, it seems like this is the content

File: spec/models/style_guide/coffee_script_spec.rb [hunk 10]
```diff
@@ -179,12 +170,21 @@ def violations_in(content, repository_owner_name: "ralph")
...
-    def build_file(content)
-      line = double("Line", content: "blah", number: 1, patch_position: 2)
-      double(:file, content: content, filename: "test.coffee", line_at: line)
+    def build_file(content, filename = "test.coffee")
+      line = double(
+        "Line",
+        changed?: true,
+        content: "blah",
+        number: 1,
+        patch_position: 2
```
Comment (teoljungberg) [inline spec/models/style_guide/coffee_script_spec.rb]: How about adding a trailing comma?

File: spec/models/style_guide/coffee_script_spec.rb [hunk 11]
```diff
@@ -179,12 +170,21 @@ def violations_in(content, repository_owner_name: "ralph")
...
+      )
+      double(:file, content: content, filename: filename, line_at: line)
```
Comment (teoljungberg) [inline spec/models/style_guide/coffee_script_spec.rb]: Can we use strings as the name of the double for consistency?

Comment (teoljungberg) [inline spec/models/style_guide/coffee_script_spec.rb]: Can this be split up into several lines for consistency with the `Line` double?

File: spec/models/style_guide/java_script_spec.rb [hunk 12]
```diff
@@ -3,26 +3,32 @@
 describe StyleGuide::JavaScript do
   include ConfigurationHelper
 
-  describe "#violations_in_file" do
+  describe "#file_review" do
+    it "returns completed file review" do
+      repo_config = double("RepoConfig", enabled_for?: true, for: {})
```
Comment (teoljungberg) [inline spec/models/style_guide/java_script_spec.rb]: Could we do the same with the repo config factory method here?

File: spec/models/style_guide/ruby_spec.rb [hunk 13]
```diff
@@ -1,8 +1,17 @@
 require "rails_helper"
 
-describe StyleGuide::Ruby, "#violations_in_file" do
+describe StyleGuide::Ruby, "#file_review" do
```
Comment (teoljungberg) [inline spec/models/style_guide/ruby_spec.rb]: Can we move the `describe "#file_review"` into a line of it's own?

File: spec/models/style_guide/ruby_spec.rb [hunk 14]
```diff
@@ -1,8 +1,17 @@
...
   include ConfigurationHelper
 
+  it "returns completed file review" do
```
Comment (teoljungberg) [inline spec/models/style_guide/ruby_spec.rb]: `a completed`?

File: spec/models/style_guide/scss_spec.rb [hunk 15]
```diff
@@ -1,7 +1,15 @@
 require "rails_helper"
 
 describe StyleGuide::Scss do
-  describe "#violations_in_file" do
+  describe "#file_review" do
+    it "returns completed file review" do
```
Comment (teoljungberg) [inline spec/models/style_guide/scss_spec.rb]: `a completed`?

File: spec/models/violation_spec.rb [hunk 16]
```diff
@@ -1,12 +1,7 @@
 require "rails_helper"
 
 describe Violation do
-  it { should belong_to(:build) }
-
-  describe "validations" do
-    it { should validate_presence_of(:build) }
-    it { should validate_presence_of(:filename) }
-  end
+  it { should belong_to(:file_review) }
```
Comment (teoljungberg) [inline spec/models/violation_spec.rb]: How about adding so we validate the presence of the file_review aswell?

File: spec/services/build_report_spec.rb [hunk 17]
```diff
@@ -6,36 +6,31 @@
       it "comments a maximum number of times" do
         stub_const("::BuildReport::MAX_COMMENTS", 1)
         commenter = stubbed_commenter(comment_on_violations: true)
-        build = create(:build, violations: build_list(:violation, 2))
+        file_review = create(
+          :file_review,
+          violations: build_list(:violation, 2)
```
Comment (teoljungberg) [inline spec/services/build_report_spec.rb]: Could we add a trailing comma here?

File: app/models/style_guide/coffee_script.rb (hunk 1 above)
Comment (gylaz) [inline app/models/style_guide/coffee_script.rb]: Huh, really? TIL. Is that a AR thing?

File: app/models/style_guide/coffee_script.rb (hunk 2 above)
Comment (gylaz) [inline app/models/style_guide/coffee_script.rb]: I agree that `each` is probably what should be used. `map` communicates a certain intent that we're not using here. IMO, "for debugging purposes" is YAGNI.

File: db/migrate/20150526214750_associate_violations_with_file_reviews.rb (hunk 4 above)
Comment (gylaz) [inline db/migrate/20150526214750_associate_violations_with_file_reviews.rb]: Good catch!

File: app/models/style_guide/coffee_script.rb (hunk 1 above)
Comment (thorncp) [inline app/models/style_guide/coffee_script.rb]: [Yup](http://api.rubyonrails.org/classes/ActiveRecord/Base.html#class-ActiveRecord::Base-label-Creation)

File: app/models/style_guide/coffee_script.rb (hunk 2 above)
Comment (thorncp) [inline app/models/style_guide/coffee_script.rb]: It would just be an array of `true` values anyway.

File: spec/models/violation_spec.rb (hunk 16 above)
Comment (gylaz) [inline spec/models/violation_spec.rb]: It's already validated at the DB layer. There is no value of adding a Rails validation because users aren't directly creating validations.

File: spec/models/style_guide/ruby_spec.rb (hunk 13 above)
Comment (gylaz) [inline spec/models/style_guide/ruby_spec.rb]: I think, I'm going to hold off on fixing this in this PR, as it will cause a lot more diff noise.

File: spec/models/style_checker_spec.rb (hunk 7 above)
Comment (gylaz) [inline spec/models/style_checker_spec.rb]: This would cause everything to shift, adding more diff noise. I agree in general, but not in this PR :smile:

File: spec/models/style_guide/coffee_script_spec.rb (hunk 11 above)
Comment (gylaz) [inline spec/models/style_guide/coffee_script_spec.rb]: If it fits on one line, I prefer it to keep it like that. I hear the consistency argument, but then this method length would grow -- we could go about extracting that, but this is just test helpers.

File: spec/models/style_checker_spec.rb (hunk 7 above)
Comment (teoljungberg) [inline spec/models/style_checker_spec.rb]: You can remove the diff noise by adding `?w=1` in the url. But maybe it's a better fit for another PR, but my thinking if it's moved forward to another PR it will never be dealt with

File: app/models/style_guide/ruby.rb [hunk 18]
```diff
@@ -3,26 +3,26 @@ module StyleGuide
   class Ruby < Base
     DEFAULT_CONFIG_FILENAME = "ruby.yml"
 
-    def violations_in_file(file)
+    def file_review(file)
       if config.file_to_exclude?(file.filename)
-        []
+        ExcludedFileReview.new
```
Comment (salbertson) [inline app/models/style_guide/ruby.rb]: Why not return a `FileReview` object without violations?

File: app/models/file_review.rb [hunk 19]
```diff
@@ -1,7 +1,18 @@
 class FileReview < ActiveRecord::Base
   belongs_to :build
+  has_many :violations, dependent: :destroy
```
Comment (salbertson) [inline app/models/file_review.rb]: When would file reviews be destroyed?

File: app/models/style_guide/coffee_script.rb [hunk 20]
```diff
@@ -4,18 +4,18 @@ class CoffeeScript < Base
...
-        line = file.line_at(violation["lineNumber"])
-
-        Violation.new(
-          filename: file.filename,
-          line: line,
-          patch_position: line.patch_position,
-          line_number: violation["lineNumber"],
-          messages: [violation["message"]]
-        )
+    def file_review(file)
```
Comment (salbertson) [inline app/models/style_guide/coffee_script.rb]: Ha, busted! Is this check useful?

File: app/models/style_guide/unsupported.rb [hunk 21]
```diff
@@ -1,8 +1,8 @@
 # Returns empty set of violations.
 module StyleGuide
   class Unsupported < Base
-    def violations_in_file(_)
-      []
+    def file_review(_)
+      ExcludedFileReview.new
```
Comment (salbertson) [inline app/models/style_guide/unsupported.rb]: Same here, could we use a `FileReview` object that doesn't have any violations?

File: db/migrate/20150526214750_associate_violations_with_file_reviews.rb [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+  private
+
+  def associate_violations_with_file_reviews
+    Violation.find_each do |violation|
+      attributes = {
+        build_id: violation.build_id,
+        filename: violation.filename,
+      }
+
+      file_review = FileReview.find_or_create_by!(attributes) do |review|
```
Comment (salbertson) [inline db/migrate/20150526214750_associate_violations_with_file_reviews.rb]: Are there `FileReview` records to find?

File: spec/factories.rb [hunk 23]
```diff
@@ -4,18 +4,16 @@
...
   factory :file_review do
-    build
-
     trait :completed do
       completed_at Time.zone.now
     end
+
+    build
+
+    filename "the_thing.rb"
```
Comment (salbertson) [inline spec/factories.rb]: Should we use a unique string for filename?

File: app/models/style_checker.rb [hunk 24]
```diff
@@ -7,20 +7,16 @@ def initialize(pull_request)
     @style_guides = {}
   end
 
-  def violations
-    @violations ||= Violations.new.push(*violations_in_checked_files).to_a
```
Comment (salbertson) [inline app/models/style_checker.rb]: Is `Violations` no longer used?

File: app/models/file_review.rb [hunk 25]
```diff
@@ -1,7 +1,18 @@
...
 
-  validates :build, presence: :true
+  def build_violation(line, line_number, message)
```
Comment (salbertson) [inline app/models/file_review.rb]: Can we send `line` and get `line_number` from that object instead of passing two arguments?

File: app/models/style_checker.rb [hunk 26]
```diff
@@ -7,20 +7,16 @@ def initialize(pull_request)
...
+  def file_reviews
+    files_to_check.map do |file|
```
Comment (teoljungberg) [inline app/models/style_checker.rb]: How about using `flat_map` to avoid `nil` values?

**Comment by teoljungberg** [comment]
@gylaz I weighed in and updated the PR with the newly merged HAML support

File: app/models/style_guide/coffee_script.rb (hunk 20 above)
Comment (gylaz) [inline app/models/style_guide/coffee_script.rb]: It hasn't been too bad yet, and causes to extract smaller methods. However, the comment will almost never collapse, since this line isn't likely to change.

File: app/models/style_guide/scss.rb [hunk 27]
```diff
@@ -2,31 +2,36 @@ module StyleGuide
...
-        []
+        ExcludedFileReview.new
       else
+        perform_file_review(file)
+      end
+    end
+
+    private
+
+    def perform_file_review(file)
```
Comment (gylaz) [inline app/models/style_guide/scss.rb]: I'm not sure this one is helpful at all. I think it's best humans review and judge complexity.

File: app/models/style_guide/ruby.rb (hunk 18 above)
Comment (gylaz) [inline app/models/style_guide/ruby.rb]: Great idea.

File: app/models/file_review.rb (hunk 19 above)
Comment (gylaz) [inline app/models/file_review.rb]: Likely never, but they could be manually. Are you opposed to having the `dependent: :destroy` here?

File: db/migrate/20150526214750_associate_violations_with_file_reviews.rb (hunk 22 above)
Comment (gylaz) [inline db/migrate/20150526214750_associate_violations_with_file_reviews.rb]: No. We'll create one for the first violation in a file, then re-use that record for the other violations.

File: spec/factories.rb (hunk 23 above)
Comment (gylaz) [inline spec/factories.rb]: The same filename could be reviewed multiple times. Is there a particular benefit to have it be unique?

File: app/models/style_checker.rb (hunk 26 above)
Comment (gylaz) [inline app/models/style_checker.rb]: `flat_map` doesn't get rid of `nil`s. Are you thinking of `compact`?

There should never be a `nil` returned from `file_review`. Am I misunderstanding?

File: app/models/file_review.rb [hunk 28]
```diff
@@ -10,4 +21,11 @@ def completed?
   def running?
     !completed?
   end
+
+  private
+
+  def find_or_build_violation(line)
+    violations.detect { |violation| violation.line_number == line.number } ||
+      violations.build(line_number: line.number)
```
Comment (thorncp) [inline app/models/file_review.rb]: Could this be `violations.find_or_initialize_by(line_number: line.number)`?

File: app/models/style_guide/coffee_script.rb (hunk 20 above)
Comment (thorncp) [inline app/models/style_guide/coffee_script.rb]: Yeah, I think the comment not collapsing is the biggest drawback. I like the check itself.

**Comment by thorncp** [comment]
LGTM :+1:  I like how this cleans up the style guide classes.

File: app/models/file_review.rb (hunk 28 above)
Comment (gylaz) [inline app/models/file_review.rb]: Good idea, but unfortunately, I don't think so. The reason this is using `detect` is because the violations aren't in the database -- they are being built in memory as the file is being reviewed. I've added a test in 9c5f3ed to ensure this.

Comment (thorncp) [inline app/models/file_review.rb]: Ah, yeah good point :+1:

File: app/models/style_checker.rb (hunk 26 above)
Comment (teoljungberg) [inline app/models/style_checker.rb]: You're right, thanks.

File: app/models/style_guide/coffee_script.rb (hunk 20 above)
Comment (salbertson) [inline app/models/style_guide/coffee_script.rb]: Ah, didn't think about collapsing, that's a bummer.

File: app/models/violations.rb (hunk 3 above)
Comment (salbertson) [inline app/models/violations.rb]: Noice

File: app/models/style_guide/scss.rb (hunk 27 above)
Comment (salbertson) [inline app/models/style_guide/scss.rb]: Yea, go ahead and disable if it isn't useful.

File: app/models/file_review.rb (hunk 19 above)
Comment (salbertson) [inline app/models/file_review.rb]: Yea, it suggests that deleting could happen in the app. I think if they are deleted manually the person should be deleting associated records, but I'm not convinced violations should ever be deleted, even manually.

File: db/migrate/20150526214750_associate_violations_with_file_reviews.rb (hunk 22 above)
Comment (salbertson) [inline db/migrate/20150526214750_associate_violations_with_file_reviews.rb]: Ok, so should we be creating instead of finding or creating?

File: spec/factories.rb (hunk 23 above)
Comment (salbertson) [inline spec/factories.rb]: No particular benefit, that makes sense.

File: app/models/style_guide/ruby.rb (hunk 18 above)
Comment (salbertson) [inline app/models/style_guide/ruby.rb]: Awesome, I like it.

File: app/models/file_review.rb (hunk 19 above)
Comment (gylaz) [inline app/models/file_review.rb]: FWIW, the foreign_key on the db will ensure cascade deleting, and I would keep that for data integrity. I'll remove it from the model.

File: db/migrate/20150526214750_associate_violations_with_file_reviews.rb (hunk 22 above)
Comment (gylaz) [inline db/migrate/20150526214750_associate_violations_with_file_reviews.rb]: Perhaps I don't follow your logic, but here's how I'm thinking about this:

* Get the first violation
* Create a file review record for it
* Get the second violation which happens to be for the same file and build as the first one
* Find the existing file review record and associate the two

Comment (gylaz) [inline db/migrate/20150526214750_associate_violations_with_file_reviews.rb]: Basically, we don't want to create a new `file_review` record for each existing violation.

Comment (salbertson) [inline db/migrate/20150526214750_associate_violations_with_file_reviews.rb]: Ah, that makes sense.

**Comment by salbertson** [comment]
Looks good to me. :+1:

**Comment by thorncp** [comment]
Updated the `up` migration to use SQL in d3d3168de95ab0d1f833ed151e05c52584e77ca2. Definitely needs :eyes: on it to make sure the data is correct.

I pulled the production database locally and ran this migration. After an hour and fifteen minutes I killed the killed the process. With SQL version it executes in about three minutes.
